### DISCUSSION TITLE
Pattern to handle large projects with multiple version deps

### DISCUSSION THREAD

**Discussion by jwinarske** [meta-rust/meta-rust#348] [Q&A]
I'm working on a recipe for [firecracker](https://github.com/jwinarske/meta-microvm/blob/main/recipes-containers/firecracker/firecracker_0.24.3.bb).  It's not yet building.

The stack has multiple version requirements between all the projects.

What is the recommended approach for this?  Adjust the stack so all the dependencies have the same version?  Most are off by x.x.1.

It really illustrates one of Rust's less glamorous traits (I call it the Rust [code version] flea market), and I'm surprised there isn't a tool to help optimize for this.  Maybe there is and I just haven't found it?

**Comment by srwalter** [comment]
If it builds outside of bitbake, I would expect it to build with cargo-bitbake as well.  What's the specific failure?

I do tend to agree that things will go better for you if you can unify dependency versions as much as possible.  I think the only way you would get multiple x.y.1, x.y.2, etc is if crates are specifying exact patch versions in their Cargo.toml, instead of just x.y.  That should let cargo automatically unify versions in a lot of cases.  Of course I would expect the generated output to be smaller with unified versions, as well.

**Reply by jwinarske** [reply]
I see.  Then one approach would be to patch toml files to trim trailing version specifier.  I think as a matter of practice it makes sense until you have a release build, which should have explicit versioning.  Then there is the case of git submodules.  These are no guarantees on what versions they use.  

In below recipe I adopted use of a tarball to avoid the ci git submodules.  I will move back to git once it's clear how to handle many version of the same module.

Current build failure:  https://gist.github.com/jwinarske/7c0316e723d9013d45f78e17d626bbe9
I added redundant versions in recipe to work around it complaining about not finding version match.  Seems like the last occurring version referenced is overwriting the prior.

https://github.com/jwinarske/meta-microvm/blob/main/recipes-containers/firecracker/firecracker_0.24.3.bb

**Reply by srwalter** [reply]
Another possibility might be using the [patch] section in your Cargo.toml to override dependencies: https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-patch-section

**Reply by jwinarske** [reply]
I'll give that a go.  Thanks

**Reply by jwinarske** [reply]
End result
https://github.com/jwinarske/meta-microvm/tree/main/recipes-containers/firecracker

**Reply by srwalter** [reply]
Glad you got it working!  I would recommend that you use the result of cargo-bitbake verbatim if at all possible.  Customizations can be made in the .inc file that the generated recipe includes.  This allows you to re-run cargo-bitbake whenever you need to (when dependencies change, for example)

**Reply by jwinarske** [reply]
I hear you.  I realize that is preferred format for rolling versions.  I don't believe that works out of the box.  I'll try it out later.

**Reply by jwinarske** [reply]
I refactored it and I put PR into meta-aws, as well as upstream for changes to support cargo-bitbake:
https://github.com/aws/meta-aws/pull/128
https://github.com/firecracker-microvm/firecracker/pull/2577

Now cargo-bitbake just needs to sort out how to automate license detection :)

**Reply by srwalter** [reply]
Looks good!
